Install python-apt package if necessary

wtanaka opened the pull request:
This solves an issue where under certain configurations, ansible would
emit this error message:

```
msg: Could not import python modules: apt, apt_pkg. Please install
python-apt package.
```

wtanaka committed:
Install python-apt package if necessary

This solves an issue where under certain configurations, ansible would
emit this error message:

```
msg: Could not import python modules: apt, apt_pkg. Please install
python-apt package.
```

ahelal commented:
Thank you,

I think this might be problematic. It will break on non ubuntu system. If you're using an old Ansible can't you use `raw: "apt-get install python-apt -y"` in your play ?

wtanaka commented:
On a non Ubuntu system, my intention is for the `-x` check to short circuit the execution of apt-get, and thus I expect this patch to maintain the previous behavior/error message of this code.

ahelal commented:
That's cool, but I still see a case that this might break. If apt-get update was never run on the target, Then the **_apt-get install**_ will fail. So If you do decide todo `apt-get update` that would be costy and not elegant.
So my suggestion would be 
A) ignore the issue and delegate to your playbook or role. I usually run an apt-role in the start that takes care of apt that case could be easily added.
B) Check the host ansible version than 1- python-apt 2- update 3- install 
C) Upgrade Ansible :) 
D) If you have another idea I am open

wtanaka commented:
that ^^^ was just a rebase with no other changes, I'll take a crack at option B

wtanaka committed:
Validate config on demand (and memoize it)

wtanaka committed:
Extract method machine_name

wtanaka committed:
Extract method command_env

wtanaka committed:
Extract method options()

wtanaka committed:
Extract method command()

wtanaka committed:
Remove dead code compile_config

wtanaka committed:
Only install python-apt for ansible < 1.6.0

Starting in Ansible 1.6.0, ansible does this on its own, see ansible
commit e4ad97b9186338c5aa683716d07ca305fd815185

wtanaka committed:
Extract method instance_connection_option()

Previously, prepare_inventory had the side effect of also setting
@instance_connection_option, which prevented using
@instance_connection_option prior to calling prepare_inventory

wtanaka committed:
Run apt-get update prior to apt-get install

ahelal commented:
Thank you.

I am currently on vacation will have a look at it late this week.

Regards
Adham

mcamou commented on lib/kitchen/provisioner/ansible_push.rb:
```diff
@@ -39,10 +40,108 @@ class AnsiblePush < Base
       # For tests disable if not needed
       default_config :chef_bootstrap_url, "https://www.getchef.com/chef/install.sh"
 
+      attr_reader :conf
```
Why have an `attr_reader :conf` if you already have a method called `conf`?

wtanaka replied:
Syntactic sugar, I can remove it if you prefer; code would look like 

conf()[:some_key]

?

ahelal replied:
but isn't it true that in Ruby when calling a no-argument method you don't have to specify parentheses? For example:

```
irb(main):004:0> def x
irb(main):005:1> { :a => 1, :b => 2 }
irb(main):006:1> end
=> :x
irb(main):007:0> x[:a]
=> 1
```

To my understanding  `attr_reader conf``and``attr_reader :machine_name` is redundant

wtanaka replied:
@ahelal Removed and updated pull request

mcamou commented on lib/kitchen/provisioner/ansible_push.rb:
```diff
@@ -39,10 +40,108 @@ class AnsiblePush < Base
+      def conf
+        return @validated_config if defined? @validated_config
+
+        if !config[:playbook]
```
could we change this to `config[:playbook].nil?` ?

mcamou commented on lib/kitchen/provisioner/ansible_push.rb:
```diff
@@ -39,10 +40,108 @@ class AnsiblePush < Base
+          raise "playbook '%s' could not be found. Please check path" % config[:playbook]
+        end
+
+        if config[:vault_password_file] and !File.exist?(config[:vault_password_file])
```
Use `.nil?` and `unless` in a postfix position

mcamou commented on lib/kitchen/provisioner/ansible_push.rb:
```diff
@@ -39,10 +40,108 @@ class AnsiblePush < Base
+        end
+
+        # Validate that extra_vars is either a hash, or a path to an existing file
+        if config[:extra_vars]
```
`.nil?`

mcamou commented on lib/kitchen/provisioner/ansible_push.rb:
```diff
@@ -39,10 +40,108 @@ class AnsiblePush < Base
+              @extra_vars = '@' + extra_vars_path
+            end
+          end
+          if !extra_vars_is_valid
```
use `unless` and put in postfix position

mcamou commented on lib/kitchen/provisioner/ansible_push.rb:
```diff
@@ -50,12 +149,19 @@ def prepare_command
+        omnibus_download_dir = conf[:omnibus_cachier] ? "/tmp/vagrant-cache/omnibus_chef" : "/tmp"
+        chef_url = conf[:chef_bootstrap_url]
+
         if chef_url
```
`unless chef_url.nil?`

wtanaka replied:
I didn't touch this line either way in this pull request

wtanaka commented on lib/kitchen/provisioner/ansible_push.rb:
```diff
@@ -123,111 +251,41 @@ def exec_ansible_command(env, command, desc)
-      end
-
-      def validate_config()
-        if !config[:playbook]
```
@mcamou  all this code already existed, I just moved it around and extracted functions, it doesn't seem like making further changes of coding style would be relevant to this specific pull request?

ahelal replied:
@mcamou @wtanaka I would take care of that part.

wtanaka committed:
Remove extraneous attr_readers

wtanaka committed:
Install python-pycurl if needed

(for older versions of ansible)
